### ISSUE TITLE
f_constant_pad_nd arguments

### ISSUE THREAD

**Issue by guillaume-be** [LaurentMazare/tch-rs#296]
Hello,

I would like to use the equivalent of `torch.nn.functional.pad` (see https://pytorch.org/docs/stable/nn.functional.html#pad). The closest I could find in the library is  `f_constant_pad_nd` (the bindings to https://github.com/pytorch/pytorch/blob/22a34bcf4e5eaa348f0117c414c3dd760ec64b13/aten/src/ATen/native/ConstantPadNd.cpp#L5)

https://github.com/LaurentMazare/tch-rs/blob/50568a42a43c97003215c10375255765ec63d372/src/wrappers/tensor_fallible_generated.rs#L6269

I might be missing something, but it seems there is a mismatch between the arguments to the Pytorch/Libtorch methods and the Rust bindings.

The Pytorch method takes an array of input containing the amount of padding to add per dimension (it needs to be even, and at most of size 2x the tensor rank). This describes the amount of padding to add before and after for each dimension.

It also takes an additional input, the padding value to pad with. This additional input seems to be missing from the Rust bindings. 
Could you please point me in the right direction? 

Thank you!

**Comment by LaurentMazare** [comment]
Hello,
Indeed there is a limitation when generating the wrappers: optional arguments are not handled so we always use the default values for these. If you look in the [Declarations.yaml](https://raw.githubusercontent.com/LaurentMazare/tch-rs/master/third_party/pytorch/Declarations-v1.7.0.yaml) file, the 'value' parameter of 'constant_pad_nd' is optional with a default value of 0.
The main reason for this is that default values as handled in C++/Python don't have a very idiomatic equivalent in Rust (there is the default trait but you still have to pass an argument). I guess handling default values would first require having named arguments, which would be pretty nice in Rust, for which there have been some discussions but nothing concrete yet.
Maybe we should tweak this crate to generate Rust arguments for arguments with default values, this clutters the function definition a bit though.
Anyway would 0 work for your use case or do you need to use a different value?

**Comment by guillaume-be** [comment]
Hello,

Thank you for the response, this is very interesting. I was navigating the bindings generation code for the first time and have seen the declarations file, but I did not realize that optional parameters were ignore from the bindings generated.

I agree that the absence of optional inputs to methods is fairly cumbersome in Rust, especially to add further capabilities while preserving backward compatibility. Waiting for a (hypothetical?) implementation of named arguments to handle optional values more smoothly, maybe an option would be to handle these methods with optional arguments the same way you implemented function overloading from C++ and Python.

For example, you converted the `arange` method into several methods called `arange`, `arange1`, `arange2`...
https://github.com/pytorch/pytorch/blob/4883d39c6fd38431bdc60e1db6402e251429b1e1/aten/src/ATen/native/TensorFactories.cpp#L70

I agree the API should remains uncluttered as possible, and the pad value of 0 should work for most (it works for me). Maybe there is a way to generate further bindings exposing additional optional parameters, e.g. 
```rust
pub fn f_constant_pad_nd(&self, pad: &[i64]) -> Result<Tensor, TchError> {
 ...
}

pub fn f_constant_pad_nd1<S: Into<Scalar>>(&self, pad: &[i64], value: S) -> Result<Tensor, TchError> {
 ...
}
```
The drawback is a multiplication of Tensor associated methods that can make the general API a bit more cluttered.
I don't need to change the value for my use-case, so this is not a must-have for me. This is also the first time I encounter a situation where I may have had the need to change this value

Thanks again for the insights!

**Comment by LaurentMazare** [comment]
For overloading we indeed generate multiple methods which was a simple way of handling it but feels somewhat unidiomatic with the method names ending with 1/2/... (maybe there could be a better approach with some into traits but the number of parameters here might not even be the same).
Something similar could be done for optional parameters as per your suggestion, one culprit is that if you want to handle each optional parameter independently you would end up with a number of methods exponential in the number of optional parameters which would be cumbersome. An alternative could be to only have two methods, "use default values for all optional arguments" vs "specify values for all optional arguments". Maybe that would be the best compromise though it would generate a large number of additional methods.
This "use default values" behaviour came up twice as far as I remember (#132, #133), it's certainly good to hear that there are more people running into this, though it's unclear at which point we should start changing this.

**Comment by guillaume-be** [comment]
Thank you for the feedback. I agree a version of the function for each combination of optional argument is not sensible, the `all default` and `no defaults` could be a good alternative, although as you mentioned it would duplicate a lot of existing methods.
I found a work-around so that I do not require setting a non-default value, so I will close this issue for now.
